File: src/flavor/rust/pspf-launcher-rs/src/metadata.rs
//! PSPF metadata structures and types
//! 
//! This module defines all the metadata structures used in PSPF packages,
//! matching the PSPF 2025 specification.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

/// Main metadata structure for a PSPF package
#[derive(Debug, Deserialize, Serialize)]
pub struct Metadata {
    pub format: String,
    pub package: PackageInfo,
    pub slots: Vec<SlotMetadata>,
    pub execution: ExecutionInfo,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub build: Option<BuildInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_validation: Option<CacheValidationInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime: Option<RuntimeInfo>,
    #[serde(default)]
    pub setup_commands: Vec<Value>,
}

/// Package information
#[derive(Debug, Deserialize, Serialize)]
pub struct PackageInfo {
    pub name: String,
    pub version: String,
}

/// Slot metadata for each data slot in the package
/// 
/// Key fields:
/// - `size`: Size as stored in the package (after any compression)
/// - `encoding`: Indicates compression type ("none", "gzip")
/// - `extract_to`: Optional subdirectory for extraction
#[derive(Debug, Deserialize, Serialize)]
pub struct SlotMetadata {
    pub index: usize,
    pub name: String,
    pub size: i64,  // Size as stored in package
    pub checksum: String,
    pub encoding: String,  // Indicates compression type
    pub purpose: String,
    pub lifecycle: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extract_to: Option<String>,  // Runtime extraction subdirectory
}

/// Execution configuration
#[derive(Debug, Deserialize, Serialize)]
pub struct ExecutionInfo {
    pub primary_slot: usize,
    pub command: String,
    #[serde(default)]
    pub environment: HashMap<String, String>,
}

/// Build information (optional)
#[derive(Debug, Deserialize, Serialize)]
pub struct BuildInfo {
    pub builder: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
}

/// Cache validation configuration
#[derive(Debug, Deserialize, Serialize)]
pub struct CacheValidationInfo {
    pub check_file: String,
    pub expected_content: String,
}

/// Runtime configuration
#[derive(Debug, Deserialize, Serialize)]
pub struct RuntimeInfo {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<RuntimeEnv>,
}

/// Runtime environment configuration
/// 
/// Operations are processed in this order:
/// 1. Analyze pass patterns (what to preserve)
/// 2. unset - Remove specified variables (except preserved)
/// 3. map - Rename variables
/// 4. set - Set specific values
/// 5. pass verification - Check required variables exist
#[derive(Debug, Deserialize, Serialize)]
pub struct RuntimeEnv {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unset: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub map: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub set: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pass: Option<Vec<String>>,
}

File: src/flavor-go/pkg/psp/format_2025/metadata.go
package format_2025

type Metadata struct {
	Format          string               `json:"format"`
	FormatVersion   string               `json:"format_version"`
	Package         PackageInfo          `json:"package"`
	CacheValidation *CacheValidationInfo `json:"cache_validation,omitempty"`
	SetupCommands   []interface{}        `json:"setup_commands,omitempty"`
	Slots           []SlotMetadata       `json:"slots"`
	Execution       *ExecutionInfo       `json:"execution,omitempty"`
	Runtime         *RuntimeInfo         `json:"runtime,omitempty"`
	Verification    *VerificationInfo    `json:"verification,omitempty"`
	Build           *BuildInfo           `json:"build,omitempty"`
	Launcher        *LauncherInfo        `json:"launcher,omitempty"`
	Compatibility   *CompatibilityInfo   `json:"compatibility,omitempty"`
	Workenv         *WorkenvInfo         `json:"workenv,omitempty"`
}

type PackageInfo struct {
	Name        string `json:"name"`
	Version     string `json:"version"`
	Description string `json:"description"`
}

type CacheValidationInfo struct {
	CheckFile       string `json:"check_file"`
	ExpectedContent string `json:"expected_content,omitempty"`
}

type ExecutionInfo struct {
	PrimarySlot int               `json:"primary_slot"`
	Command     string            `json:"command"`
	Environment map[string]string `json:"environment"`
}

type RuntimeInfo struct {
	Env map[string]interface{} `json:"env,omitempty"`
}

type WorkenvInfo struct {
	Directories []DirectorySpec   `json:"directories,omitempty"`
	Env         map[string]string `json:"env,omitempty"`
}

type DirectorySpec struct {
	Path string `json:"path"`
	Mode string `json:"mode,omitempty"`
}

type VerificationInfo struct {
	IntegritySeal       IntegritySealInfo    `json:"integrity_seal"`
	Signed              bool                 `json:"signed"`
	RequireVerification bool                 `json:"require_verification"`
	TrustSignatures     *TrustSignaturesInfo `json:"trust_signatures,omitempty"`
}

type IntegritySealInfo struct {
	Required  bool   `json:"required"`
	Algorithm string `json:"algorithm"`
}

type TrustSignaturesInfo struct {
	Required bool         `json:"required"`
	Signers  []SignerInfo `json:"signers,omitempty"`
}

type SignerInfo struct {
	Name      string `json:"name"`
	KeyID     string `json:"key_id"`
	Algorithm string `json:"algorithm"`
}

type BuildInfo struct {
	Tool          string       `json:"tool"`
	ToolVersion   string       `json:"tool_version"`
	Timestamp     string       `json:"timestamp"`
	Deterministic bool         `json:"deterministic"`
	Platform      PlatformInfo `json:"platform"`
}

type PlatformInfo struct {
	OS   string `json:"os"`
	Arch string `json:"arch"`
	Host string `json:"host"`
}

type LauncherInfo struct {
	Tool         string   `json:"tool"`
	ToolVersion  string   `json:"tool_version"`
	Size         int64    `json:"size"`
	Checksum     string   `json:"checksum"`
	Capabilities []string `json:"capabilities"`
}

type CompatibilityInfo struct {
	MinFormatVersion string   `json:"min_format_version"`
	Features         []string `json:"features"`
}
